wordpress-mobile/WordPress-iOS
# Tools/update release tools

This PR adds some scripts that handle a part of the code freeze process.
This block of scripts are introduced together because it's easier to test them this way, since every script takes over when the previous ends.

###To Test
Testing can be a bit tricky as some of these operations are to be done only during an actual release. Anyway, all of them can be reverted, so it's possible to test it. Just remember to go back and delete all the changes. 
The following instructions are detailed for the current scenario, with 10.6 code frozen and 10.7 in development. The version numbers shall be updated if the test is performed under a different scenario.

1. Setup the GitHub access;
2. Put a fake 10.6 release tag in the repository;
3. Test the code freeze script: 
3.a `cd Scripts` and `bundle exec fastlane code_freeze`;
3.b when asked, verify the it has picked the correct current version and answer y;
3.c When it finishes, verify that:
- there's a new release/10.7 branch in the repo;
- the version in the xcconfig, DeliverFile and download_metadata.swift files has been updated;
- the new branch has been pushed to GitHub, with all the changes and tags;
- the right protection rules are set for the branch;
- the 10.7 has the frozen tag;
- verify that the ~/wpios_prs_list_10.6_10.7.txt file has been generated in your home with the list of merged PRs.


4. Merge the test branch into the new release branch, in order to have the scripts available: `git checkout release/10.7` `git merge tools/update-release-tools`

5. Test the new beta release script:
5.a `cd Scripts` and `bundle exec fastlane new_beta_release base_version:10.7`
5.b When it finishes, verify that:
- the version in the xcconfig files have been updated;
-  the changes have been pushed to GitHub;
- the version has been tagged;
NOTE: the script will report an error, because of the way we tag the internal version. We use the today date as the last version number for the internal version, so, as we are simulating two builds in the same day, the last tag will fail.
This is expected and related to the actual process (and something that is not likely to happen in the reality).

6. Test the final release script:
6.a `git checkout release/10.7` and `cd Scripts` and `bundle exec fastlane finalize_release`
6.b When it finishes, verify that:
- the localisation files in the Resources folder have been updated;
- the metadata localisation in fastlane/metadata have been updated;
- the version has been tagged with 10.7 and the 10.7.0.x tags have been removed.

7. Test the hot fix script:
7.a `cd Scripts` and `bundle exec fastlane new_hotfix_release version:10.5.2`
7.b when asked, verify that it has picked the correct current version and answer y;
7.c when it finishes, verify that:
- the new release/10.5.2 branch has been created;
- the version in the xcconfig, DeliverFile and download_metadata.swift files has been updated;
- the changes have been pushed to GitHub;
- the version has been tagged;


At the end of the test:

- remove branch protection from GitHub and delete the branch;
- remove the frozen tag from the GitHub milestone;
- delete the release branch, both local and on GitHub
- delete the fake 10.6 tag you added in point 2 of the test.
- delete the 10.7 tag the script added
- delete the 10.5.2.0 tag
